### ISSUE TITLE
make `Value()` of a Decimal function 

### ISSUE THREAD

**Issue by bubenkoff** [shopspring/decimal#36]
Currently we have this

```golang
// Value implements the driver.Valuer interface for database serialisation.
func (d *Decimal) Value() (driver.Value, error) {
	if d == nil {
		return nil, nil
	}
	return d.String(), nil
}
```

while it's unclear how can it be that decimal is nil, and why should it be a function of a pointer at all?

https://github.com/jinzhu/gorm/issues/1250#issuecomment-259585820

**Comment by msmarino** [comment]
I assume this is to allow OPTIONAL ```Decimal``` values as a pointer can be ```nil```.

However, it also forces SQL persisted structs with ```Decimal```s  to declare them as pointers.

Alternatively, in my own projects I left the above code as
```golang
func (d Decimal) Value() (driver.Value, error) {
```
(not a pointer) which functions as expected.

You can also add
```golang
// NullDecimal represents a Decimal that may be null.
// NullDecimal implements the Scanner interface so
// it can be used as a scan destination:
//
//  var s NullDecimal
//  err := db.QueryRow("SELECT name FROM foo WHERE id=?", id).Scan(&s)
//  ...
//  if s.Valid {
//     // use s.Decimal
//  } else {
//     // NULL value
//  }
//
type NullDecimal struct {
	Decimal
	Valid bool // Valid is true if Decimal is not NULL
}

// Scan implements the Scanner interface.
func (nd *NullDecimal) Scan(value interface{}) error {
	if value == nil {
		nd.Decimal, nd.Valid = New(0, 0), false
		return nil
	}
	nd.Valid = true
	return nd.Scan(value)
}

// Value implements the driver Valuer interface.
func (nd NullDecimal) Value() (driver.Value, error) {
	if !nd.Valid {
		return nil, nil
	}
	return nd.Decimal, nil
}
```
(shamelessly adapted from ```sql.NullString```) for optional values stored as ```NULL``` if you should require them.
__This latter code is _not_ tested.__

**Comment by bubenkoff** [comment]
new `NullDecimal` type makes total sense to add to this package

**Comment by james-maloney** [comment]
This is a somewhat recent change https://github.com/shopspring/decimal/commit/54efccb61ffe0a31b6a1908bdc8a35f491da01ea that is not backwards compatible.

**Comment by msmarino** [comment]
Pull Request #32 implements a full solution, undoing the mentioned breaking change AND creating a fully tested NullDecimal.

Thanks @craigjackson!

I hope this gets merged soon!

**Comment by sambengtson** [comment]
Seems like quite a few of us are using this project on a regular basis.  Would it make sense to start directing our PR's to a fork so we can get some of this stuff merged a little quicker?

**Comment by cressie176** [comment]
FYI  https://github.com/shopspring/decimal/pull/32 caused us some head scratching as it broke our code which was using *decimal.Decimal to persist optional decimals as null. After upgrading we started seeing...

```
panic: value method decimal.Decimal.Value called using nil *Decimal pointer
```

It was fixed by upgrading to go 1.8 due to https://github.com/golang/go/commit/0ce1d79a6a771f7449ec493b993ed2a720917870. Guessing you are already aware but it caught us by surprise, so posting here in case others come looking.

Otherwise thanks for decimal. We're finding it really useful.
